### PR TITLE
[BH-1453] Fix audio on the simulator

### PR DESCRIPTION
**Pull request by swichu91**
**Description**

Replaced PortAudio library with PulseAudio.
Fixed minor issues with the WorkerQueue module.

- [x] Complies with our guidelines for contributions
- [x] Has unit tests if possible.
- [x] Has documentation updated

### CODE REVIEW

File: config/bootstrap_config [hunk 1]
```diff
@@ -41,7 +41,7 @@ INSTALL_PACKAGES="
         mtools \
         ninja-build \
         pkg-config \
-        portaudio19-dev \
+        pulseaudio \
```
Comment (SP2FET) [inline config/bootstrap_config:44]: please add libpulse-dev and libpulse0

File: module-audio/board/linux/CMakeLists.txt [hunk 2]
```diff
@@ -4,10 +4,12 @@
...
 	LinuxAudioPlatform.cpp
 	LinuxAudioDevice.cpp
+		PulseAudioWrapper.cpp
 )
 
 add_library(${AUDIO_BOARD_LIBRARY} STATIC ${AUDIO_LINUX_SOURCES})
 target_include_directories(${AUDIO_BOARD_LIBRARY} INTERFACE ${CMAKE_CURRENT_SOURCE_DIR})
 target_link_libraries(${AUDIO_BOARD_LIBRARY}
     module-os
+		pulse
```
Comment (SP2FET) [inline module-audio/board/linux/CMakeLists.txt:14]: wrong formatting?

**Review by SP2FET — APPROVED** [review]

File: module-audio/board/linux/PulseAudioWrapper.cpp [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+#include <iostream>
+#include <utility>
+
+namespace audio
+{
+    PulseAudioWrapper::PulseAudioWrapper(WriteCallback write_cb, AudioFormat audio_format)
+        : write_cb{std::move(write_cb)}, audio_format(audio_format)
+    {
+        /// PulseAudio requests at least SampleRate bytes of data to start playback
+        /// Due to the how module-audio's streams operates we can load a bit more data to the cache than needed.
```
Comment (mjan-dev) [inline module-audio/board/linux/PulseAudioWrapper.cpp:15]:
```suggestion
        /// Due to how module-audio's streams operates we can load a bit more data to the cache than needed.
```

File: module-audio/board/linux/PulseAudioWrapper.cpp [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+        /// Doubling the buffer size is cheap and simple solution.
+        cache.reserve(audio_format.getSampleRate() * 2);
```
Comment (mjan-dev) [inline module-audio/board/linux/PulseAudioWrapper.cpp:17]: :+1:

File: module-audio/board/linux/PulseAudioWrapper.cpp [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+        default:
+            fprintf(stderr, "Connection failure: %s\n", pa_strerror(pa_context_errno(context)));
+            quit(1);
+        }
+    }
+
+    void PulseAudioWrapper::stream_write_cb(size_t length)
+    {
+        write_cb(length);
+    }
```
Comment (mjan-dev) [inline module-audio/board/linux/PulseAudioWrapper.cpp:110]: minor: missing newline

File: module-audio/board/linux/PulseAudioWrapper.cpp [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+        cache_pos += span.dataSize;
+    }
+    void PulseAudioWrapper::consume()
+    {
+        if (pa_stream_write(stream, cache.data(), cache_pos, NULL, 0, PA_SEEK_RELATIVE) < 0) {
+            fprintf(stderr, "pa_stream_write() failed\n");
+            return;
+        }
+        cache_pos = 0;
+    }
```
Comment (mjan-dev) [inline module-audio/board/linux/PulseAudioWrapper.cpp:149]: minor: missing newlines between methods

File: module-audio/board/linux/PulseAudioWrapper.cpp [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+        if (context != nullptr) {
+            pa_context_unref(context);
+            context = nullptr;
+        }
+
+        if (mainloop != nullptr) {
+            pa_mainloop_free(mainloop);
+            mainloop     = nullptr;
+            mainloop_api = nullptr;
```
Comment (mjan-dev) [inline module-audio/board/linux/PulseAudioWrapper.cpp:129]: is it possible to mainloop_api to not be freed here? or does pa_mainloop_free does that as well?

File: module-audio/board/linux/PulseAudioWrapper.hpp [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        void *worker();
+        void stream_write_cb(size_t length);
+        void quit(int ret = 0);
+
+        pthread_t tid{};
+        pa_stream *stream{nullptr};
+        pa_mainloop *mainloop{nullptr};
+        pa_mainloop_api *mainloop_api{nullptr};
+        pa_context *context{nullptr};
+        pa_signal_event *signal_event{nullptr};
```
Comment (mjan-dev) [inline module-audio/board/linux/PulseAudioWrapper.hpp:39]: is it used anywhere? can't find it used in changes from this PR

**Review by mjan-dev — APPROVED** [review]
Approved with minor comments.

File: module-audio/board/linux/PulseAudioWrapper.cpp (hunk 7 above)
Comment (swichu91) [reply module-audio/board/linux/PulseAudioWrapper.cpp:129]: From the documentation, it seems pa_mainloop_get_api`does not allocate anything. It just returns a pointer to some internal abstraction layer. That's how it is done in all examples I've found.

File: module-audio/board/linux/PulseAudioWrapper.hpp (hunk 8 above)
Comment (swichu91) [reply module-audio/board/linux/PulseAudioWrapper.hpp:39]: Good point, `signal_event` is indeed not used anymore :)

**Pull request merged**